pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,313 @@
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.")
+	return cmd
+}
+
+// SetupDrain populates some fields from the factory, grabs command line
+// arguments and looks up the node using Builder
+func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
+	var err error
+	if len(args) != 1 {
+		return cmdutil.UsageError(cmd, fmt.Sprintf("USAGE: %s [flags]", cmd.Use))
+	}
+
+	if o.client, err = o.factory.Client(); err != nil {
+		return err
+	}
+
+	o.mapper, o.typer = o.factory.Object()
+
+	cmdNamespace, _, err := o.factory.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	r := o.factory.NewBuilder().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		ResourceNames("node", args[0]).
+		Do()
+
+	if err = r.Err(); err != nil {
+		return err
+	}
+
+	return r.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+		o.nodeInfo = info
+		return nil
+	})
+}
+
+// RunDrain runs the 'drain' command
+func (o *DrainOptions) RunDrain() error {
+	if err := o.RunCordon(true); err != nil {
+		return err
+	}
+
+	pods, err := o.getPodsForDeletion()
+	if err != nil {
+		return err
+	}
+
+	if err = o.deletePods(pods); err != nil {
+		return err
+	}
+	cmdutil.PrintSuccess(o.mapper, false, o.out, "node", o.nodeInfo.Name, "drained")
+	return nil
+}
+
+// getPodsForDeletion returns all the pods we're going to delete.  If there are
+// any unmanaged pods and the user didn't pass --force, we return that list in
+// an error.
+func (o *DrainOptions) getPodsForDeletion() ([]api.Pod, error) {
+	pods := []api.Pod{}
+	podList, err := o.client.Pods(api.NamespaceAll).List(api.ListOptions{FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": o.nodeInfo.Name})})
+	if err != nil {
+		return pods, err
+	}
+	unreplicatedPodNames := []string{}
+
+	for _, pod := range podList.Items {
+		_, found := pod.ObjectMeta.Annotations[types.ConfigMirrorAnnotationKey]
+		if found {
+			// Skip mirror pod
+			continue
+		}
+		replicated := false
+
+		creatorRef, found := pod.ObjectMeta.Annotations[controller.CreatedByAnnotation]
+		if found {
+			// Now verify that the specified creator actually exists.
+			var sr api.SerializedReference
+			err := api.Scheme.DecodeInto([]byte(creatorRef), &sr)
+			if err != nil {
+				return pods, err
+			}
+			if sr.Reference.Kind == "ReplicationController" {
+				rc, err := o.client.ReplicationControllers(sr.Reference.Namespace).Get(sr.Reference.Name)
+				// Assume the only reason for an error is because the RC is
+				// gone/missing, not for any other cause.  TODO(mml): something more
+				// sophisticated than this
+				if err == nil && rc != nil {
+					replicated = true
+				}
+			} else if sr.Reference.Kind == "DaemonSet" {
```

Comment:
> I think you need to wrap the string with api.Kind

the type of sr.Reference.Kind is String.  See https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1829